#This script assumes you already ran cellranger-atac pipelien from 10x genomics.

#Loading librairies.
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
library(ggplot2)

#Creating a Seurat object using the peak/cell matrix and cell metadata generated by cellranger-atac, and store the path to the fragment file on disk in the Seurat object.
counts_HC <- Read10X_h5(filename = "HC-filtered_peak_bc_matrix.h5")
metadata_HC <- read.csv(
  file = "HC-singlecell.csv",
  header = TRUE,
  row.names = 1
)

counts_SOR <- Read10X_h5(filename = "SOR-filtered_peak_bc_matrix.h5")
metadata_SOR <- read.csv(
  file = "SOR-singlecell.csv",
  header = TRUE,
  row.names = 1
)

chrom_assay_HC <- CreateChromatinAssay(
  counts = counts_HC,
  sep = c(":", "-"),
  genome = 'mm10',
  fragments = 'HC-fragments.tsv.gz',
  min.cells = 10,
  min.features = 200
)

chrom_assay_SOR <- CreateChromatinAssay(
  counts = counts_SOR,
  sep = c(":", "-"),
  genome = 'mm10',
  fragments = 'SOR-fragments.tsv.gz',
  min.cells = 10,
  min.features = 200
)

HC_seuratobject <- CreateSeuratObject(
  counts = chrom_assay_HC,
  assay = "peaks",
  project = 'HC',
  meta.data = metadata_HC
)

SOR_seuratobject <- CreateSeuratObject(
  counts = chrom_assay_SOR,
  assay = "peaks",
  project = 'SOR',
  meta.data = metadata_SOR
)

#Computing QC Metrics.
#Compute nucleosome signal score per cell.
HC_seuratobject <- NucleosomeSignal(object = HC_seuratobject)
SOR_seuratobject <- NucleosomeSignal(object = SOR_seuratobject)
#Compute TSS enrichment score per cell
HC_seuratobject <- TSSEnrichment(object = HC_seuratobject, fast = FALSE)
SOR_seuratobject <- TSSEnrichment(object = SOR_seuratobject, fast = FALSE)
#Add blacklist ratio and fraction of reads in peaks.
HC_seuratobject$pct_reads_in_peaks <- HC_seuratobject$peak_region_fragments / HC_seuratobject$passed_filters * 100
HC_seuratobject$blacklist_ratio <- HC_seuratobject$blacklist_region_fragments / HC_seuratobject$peak_region_fragments
HC_seuratobject$high.tss <- ifelse(HC_seuratobject$TSS.enrichment > 2, 'High', 'Low')
TSSPlot(HC_seuratobject, group.by = 'high.tss') + NoLegend()
SOR_seuratobject$pct_reads_in_peaks <- SOR_seuratobject$peak_region_fragments / SOR_seuratobject$passed_filters * 100
SOR_seuratobject$blacklist_ratio <- SOR_seuratobject$blacklist_region_fragments / SOR_seuratobject$peak_region_fragments
SOR_seuratobject$high.tss <- ifelse(SOR_seuratobject$TSS.enrichment > 2, 'High', 'Low')
TSSPlot(SOR_seuratobject, group.by = 'high.tss') + NoLegend()
HC_seuratobject$nucleosome_group <- ifelse(HC_seuratobject$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
table(HC_seuratobject$nucleosome_group)
FragmentHistogram(object = HC_seuratobject, group.by = 'nucleosome_group', region = 'chr1-1-10000000')
SOR_seuratobject$nucleosome_group <- ifelse(SOR_seuratobject$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
table(SOR_seuratobject$nucleosome_group)
FragmentHistogram(object = SOR_seuratobject, group.by = 'nucleosome_group', region = 'chr1-1-10000000')

VlnPlot(
  object = HC_seuratobject,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 4,
)

VlnPlot(
  object = SOR_seuratobject,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 4,
)

#Remove cells that are outliers for these QC metrics.
HC_seuratobject <- subset(
  x = HC_seuratobject,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 20000 &
    pct_reads_in_peaks > 15 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)

SOR_seuratobject <- subset(
  x = SOR_seuratobject,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 20000 &
    pct_reads_in_peaks > 15 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)

#Normalization and linear dimensional reduction.
HC_seuratobject <- FindTopFeatures(HC_seuratobject, min.cutoff = 10)
HC_seuratobject <- RunTFIDF(HC_seuratobject)
HC_seuratobject <- RunSVD(HC_seuratobject)
DepthCor(HC_seuratobject)

SOR_seuratobject <- FindTopFeatures(SOR_seuratobject, min.cutoff = 10)
SOR_seuratobject <- RunTFIDF(SOR_seuratobject)
SOR_seuratobject <- RunSVD(SOR_seuratobject)
DepthCor(SOR_seuratobject)

#Non-linear dimension reduction and clustering.
HC_seuratobject <- RunUMAP(
  object = HC_seuratobject,
  reduction = 'lsi',
  dims = 2:30
)

SOR_seuratobject <- RunUMAP(
  object = SOR_seuratobject,
  reduction = 'lsi',
  dims = 2:30
)

HC_seuratobject <- FindNeighbors(
  object = HC_seuratobject,
  reduction = 'lsi',
  dims = 2:30
)

SOR_seuratobject <- FindNeighbors(
  object = SOR_seuratobject,
  reduction = 'lsi',
  dims = 2:30
)

HC_seuratobject <- FindClusters(
  object = HC_seuratobject,
  algorithm = 3,
  resolution = 1.2,
  verbose = FALSE
)

SOR_seuratobject <- FindClusters(
  object = SOR_seuratobject,
  algorithm = 3,
  resolution = 1.2,
  verbose = FALSE
)

#Create a gene activity matrix.
gene.activitiesHC <- GeneActivity(HC_seuratobject)
gene.activitiesSOR <- GeneActivity(SOR_seuratobject)
#Add the gene activity matrix to the Seurat object as a new assay and normalize it.
HC_seuratobject[['RNA']] <- CreateAssayObject(counts = gene.activitiesHC)
SOR_seuratobject[['RNA']] <- CreateAssayObject(counts = gene.activitiesSOR)
HC_seuratobject <- NormalizeData(
  object = HC_seuratobject,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(HC_seuratobject$nCount_RNA)
)
SOR_seuratobject <- NormalizeData(
  object = SOR_seuratobject,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(SOR_seuratobject$nCount_RNA)
)

DefaultAssay(SOR_seuratobject) <- 'RNA'

#Integrating with scRNA-seq data.
#Load the pre-processed scRNA-seq data.
allen_rna <- readRDS("allen_brain.rds")
allen_rna <- FindVariableFeatures(
  object = allen_rna,
  nfeatures = 5000
)

transfer.anchors <- FindTransferAnchors(
  reference = allen_rna,
  query = HC_seuratobject,
  reduction = 'cca',
  dims = 1:40
)

predicted.labelsHC <- TransferData(
  anchorset = transfer.anchors,
  refdata = allen_rna$subclass,
  weight.reduction = HC_seuratobject[['lsi']],
  dims = 2:30
)

HC_seuratobject <- AddMetaData(object = HC_seuratobject, metadata = predicted.labelsHC)

plot1 <- DimPlot(allen_rna, group.by = 'subclass', label = TRUE, repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')
plot2 <- DimPlot(HC_seuratobject, group.by = 'predicted.id', label = TRUE, repel = TRUE) + NoLegend() + ggtitle('scATAC-seq')
plot1 + plot2

DefaultAssay(HC_seuratobject) <- 'peaks'

CoveragePlot(
  object = HC_seuratobject,
  region = c("Nr4a2"),
  extend.upstream = 1000,
  extend.downstream = 1000,
  ncol = 1,
  ymax = 36
)

DefaultAssay(SOR_seuratobject) <- 'peaks'

CoveragePlot(
  object = SOR_seuratobject,
  region = c("Nr4a2"),
  extend.upstream = 1000,
  extend.downstream = 1000,
  ncol = 1
)

HC_seuratobject$dataset <- "HC"
SOR_seuratobject$dataset <- "SOR"

combined <- merge(HC_seuratobject, SOR_seuratobject)

combined <- FindTopFeatures(combined, min.cutoff = 10)
combined <- RunTFIDF(combined)
combined <- RunSVD(combined)
combined <- RunUMAP(combined, reduction = "lsi", dims = 2:30)
p1 <- DimPlot(combined, group.by = "dataset")

integration.anchors <- FindIntegrationAnchors(
  object.list = list(HC_seuratobject, SOR_seuratobject),
  anchor.features = rownames(HC_seuratobject),
  reduction = "rlsi",
  dims = 2:30
)

integrated <- IntegrateEmbeddings(
  anchorset = integration.anchors,
  reductions = combined[["lsi"]],
  new.reduction.name = "integrated_lsi",
  dims.to.integrate = 1:30,
  k.weight = 10
)

integrated <- RunUMAP(integrated, reduction = "integrated_lsi", dims = 2:30)

#Adding gene annotations to the seurat object for the mouse genome. This will allow downstream functions to pull the gene annotation information directly from the object.
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
Annotation(integrated) <- annotations

#Find differentially accessible promoter between condition for a specific cell type.
#Change back to working with peaks instead of gene activities.
DefaultAssay(integrated) <- 'peaks'
#Formatting labels for differential analysis.
integrated$celltype.dataset <- paste(Idents(integrated), integrated$dataset, sep = "_")
integrated$celltype <- Idents(integrated)
Idents(integrated) <- "celltype.dataset"
#Perform the differential analysis in Oligodendrocytes.
da_peaks_oligo <- FindMarkers(
  object = integrated,
  ident.1 = "Oligodendrocytes_SOR",
  ident.2 = "Oligodendrocytes_HC",
  test.use = 'LR',
  latent.vars = 'nCount_peaks'
)

#Example coverage plot for the gene Sgk1.
sgk1CoveragePlot <- CoveragePlot(
  object = integrated,
  region = c("Sgk1"),
  extend.upstream = 2000,
  extend.downstream = -113000)
